components/prism-yaml.js

```diff
@@ -16,32 +16,32 @@ Prism.languages.yaml = {
 		alias: 'important'
 	},
 	'datetime': {
-		pattern: /([:\-,[{]\s*(?:![^\s]+)?[ \t]*)(?:\d{4}-\d\d?-\d\d?(?:[tT]|[ \t]+)\d\d?:\d{2}:\d{2}(?:\.\d*)?[ \t]*(?:Z|[-+]\d\d?(?::\d{2})?)?|\d{4}-\d{2}-\d{2}|\d\d?:\d{2}(?::\d{2}(?:\.\d*)?)?)(?=[ \t]*(?:$|,|]|}))/m,
+		pattern: /([:\-,[{]\s*(?:\s[!&*][^\s]+)*[ \t]*)(?:\d{4}-\d\d?-\d\d?(?:[tT]|[ \t]+)\d\d?:\d{2}:\d{2}(?:\.\d*)?[ \t]*(?:Z|[-+]\d\d?(?::\d{2})?)?|\d{4}-\d{2}-\d{2}|\d\d?:\d{2}(?::\d{2}(?:\.\d*)?)?)(?=[ \t]*(?:$|,|]|}))/m,
 		lookbehind: true,
 		alias: 'number'
 	},
 	'boolean': {
-		pattern: /([:\-,[{]\s*(?:![^\s]+)?[ \t]*)(?:true|false)[ \t]*(?=$|,|]|})/im,
+		pattern: /([:\-,[{]\s*(?:\s[!&*][^\s]+)*[ \t]*)(?:true|false)[ \t]*(?=$|,|]|})/im,
 		lookbehind: true,
 		alias: 'important'
 	},
 	'null': {
-		pattern: /([:\-,[{]\s*(?:![^\s]+)?[ \t]*)(?:null|~)[ \t]*(?=$|,|]|})/im,
+		pattern: /([:\-,[{]\s*(?:\s[!&*][^\s]+)*[ \t]*)(?:null|~)[ \t]*(?=$|,|]|})/im,
```

Comment:
Unfortunately, this won't match `foo:!!tag null` and there can be [at most 2](https://yaml.org/spec/1.2/spec.html#c-ns-properties(n,c)).

```suggestion
		pattern: /([:\-,[{]\s*(?:(?:![^\s]+|[*&][^\s[\]{},]+)[ \t]*){0,2})(?:null|~)[ \t]*(?=$|,|]|})/im,
```

Reply:
Obviously there are a lot of cases that weren't tested before :(.

Reply:
Yes, but here it's a case which was highlighted before and now isn't. We generally can't expect people to format their documents properly (that'd be too easy...), so our language definitions have to robust enough to be able to deal with situations where non-essential whitespaces are missing.

And yeah. We definitely need more tests for YAML... It's such a complex language and yet we have so few test cases for it.

Reply:
And yes, there are at most 2 but we don't check if it's max 1 anchor or alias, we don't check if it is max 1 tag, we don't check if anchor _and_ alias are given (which probably is invalid).
There are a lot of semantic cases that aren't checked here. But this is neither the goal nor even possible with regexps.

Reply:
Well, yes. The goal is just to match the language as close as possible within reasonable bounds (e.g. that regexes that are less than ~200 chars long and at least a little understandable).

That why I suggested the above pattern. It's not perfect but it's closer to the ground truth compared to the current pattern.

Reply:
Oh, `foo:!!tag null` is invalid.

Reply:
Okay, it's valid but it's a scalar:
```ruby
[10] pry(main)> YAML.parse("foo:!!tag null")
=> #<Psych::Nodes::Document:0x00007f9cb9c72c90
 @children=[#<Psych::Nodes::Scalar:0x00007f9cb9c72c18 @anchor=nil, @plain=true, @quoted=false, @style=1, @tag=nil, @value="foo:!!tag null">],
 @implicit=true,
 @implicit_end=true,
 @tag_directives=[],
 @version=[]>
[11] pry(main)> YAML.parse("foo: !!tag null")
=> #<Psych::Nodes::Document:0x00007f9cb9c92fe0
 @children=
  [#<Psych::Nodes::Mapping:0x00007f9cb9c92f90
    @anchor=nil,
    @children=
     [#<Psych::Nodes::Scalar:0x00007f9cb9c92f18 @anchor=nil, @plain=true, @quoted=false, @style=1, @tag=nil, @value="foo">,
      #<Psych::Nodes::Scalar:0x00007f9cb9c92ea0 @anchor=nil, @plain=false, @quoted=false, @style=1, @tag="tag:yaml.org,2002:tag", @value="null">],
    @implicit=true,
    @style=1,
    @tag=nil>],
 @implicit=true,
 @implicit_end=true,
 @tag_directives=[],
 @version=[]>
```

Reply:
So it doesn't have to match here.

Reply:
That's unexpected. In that case, your pattern is better.